### PR TITLE
Add workaround for Ping SAML auth requiring signing headers

### PR DESCRIPTION
**Pull request by timothyb89**
Ping's SAML auth flow requires `SigAlg` and `Signature` fields to be set as query parameters during authn requests when "Enforce Signed Authn Request" is enabled in the Ping app settings. This adds a workaround to force gosaml2 to include these parameters when it otherwise would not, if `provider: ping` is set in the connector properties.

### CODE REVIEW

File: constants.go [hunk 1]
```diff
@@ -429,6 +429,8 @@ const (
 	NetIQ = "netiq"
 	// ADFS is Microsoft Active Directory Federation Services
 	ADFS = "adfs"
+	// Ping is an identity provider.
+	Ping = "ping"
```
Comment (russjones) [inline constants.go:433]: What's the official name here?

File: lib/auth/saml.go [hunk 2]
```diff
@@ -111,7 +111,18 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
 	}
 
 	req.ID = attr.Value
-	req.RedirectURL, err = provider.BuildAuthURLFromDocument("", doc)
+
+	// workaround for Ping: Ping expects `SigAlg` and `Signature` query
```
Comment (russjones) [inline lib/auth/saml.go:115]: Capitalization and punctuation needed here. Comments are full sentences.

File: lib/auth/saml_test.go [hunk 3]
```diff
@@ -130,3 +131,84 @@ func TestEncryptedSAML(t *testing.T) {
 	require.NoError(t, err)
 	require.NotEmpty(t, assertionInfo.Assertions)
 }
+
+func TestPingSAMLWorkaround(t *testing.T) {
```
Comment (russjones) [inline lib/auth/saml_test.go:135]: Add comment to test explaining the functionality you're testing.

**Review by russjones — APPROVED** [review]

File: constants.go (hunk 1 above)
Comment (timothyb89) [reply constants.go:433]: My understanding is that the underlying platform is PingIdentity, PingOne is the cloud/SaaS hosted product, and PingFederate is the on-prem product. This workaround seems to be required in all cases. Perhaps renaming it to `PingIdentity` would be more specific?

File: lib/auth/saml.go (hunk 2 above)
Comment (timothyb89) [reply lib/auth/saml.go]: Done, thanks!

File: lib/auth/saml_test.go (hunk 3 above)
Comment (timothyb89) [reply lib/auth/saml_test.go:138]: Done, thanks!

File: lib/auth/saml.go [hunk 4]
```diff
@@ -111,7 +111,18 @@ func (a *Server) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
...
 
 	req.ID = attr.Value
-	req.RedirectURL, err = provider.BuildAuthURLFromDocument("", doc)
+
+	// Workaround for Ping: Ping expects `SigAlg` and `Signature` query
+	// parameters when "Enforce Signed Authn Request" is enabled, but gosaml2
+	// only provides these parameters when binding == BindingHttpRedirect.
+	// Luckily, BuildAuthURLRedirect sets this and is otherwise identical to
+	// the standard BuildAuthURLFromDocument.
+	if connector.GetProvider() == teleport.Ping {
```
Comment (r0mant) [inline lib/auth/saml.go:120]: Is `provider` a standard field we support in SAML connector resource spec? Should probably mention it in the docs, I couldn't find any mentions of it currently.

**Review by r0mant — APPROVED** [review]
lgtm with @russjones comments addressed and the documentation updates.

File: constants.go (hunk 1 above)
Comment (timothyb89) [reply constants.go:434]: I've clarified the name in the constant's comment but can still rename it if desired.

File: lib/auth/saml.go (hunk 4 above)
Comment (timothyb89) [reply lib/auth/saml.go:120]: It is at least mentioned in the docs regarding ADFS: https://goteleport.com/docs/enterprise/sso/ssh-adfs/#create-teleport-roles

**Comment by stevenGravy** [comment]
@timothyb89 backporting just required for 6.2 and 6.3 on this.

File: constants.go (hunk 1 above)
Comment (russjones) [reply constants.go:434]: Okay sounds good to me.

File: lib/auth/saml.go (hunk 4 above)
Comment (russjones) [reply lib/auth/saml.go:120]: @timothyb89 Yes, please create a docs PR after this is merged and add @inertial-frame as reviewer. Let's document some of these workarounds we have in-place.

**Review by russjones — APPROVED** [review]

**Comment by russjones** [comment]
@timothyb89 Looks good to me, please merge the backport to `branch/v6` and `branch/v6.2`. Backporting will get simpler soon, but for now unfortunately it has to be done to two branches.

**Comment by timothyb89** [comment]
Backport PRs:

- https://github.com/gravitational/teleport/pull/7327 (`v6.2`)
- https://github.com/gravitational/teleport/pull/7328 (`v6`)

File: lib/auth/saml.go (hunk 4 above)
Comment (timothyb89) [reply lib/auth/saml.go:120]: I've just posted a docs PR here: https://github.com/gravitational/teleport/pull/7339

**Pull request merged**